README
NAME

    API::Trello - Trello.com API Client

VERSION

    version 0.06

SYNOPSIS

        use API::Trello;
    
        my $trello = API::Trello->new(
            key        => 'KEY',
            token      => 'TOKEN',
            identifier => 'APPLICATION NAME',
        );
    
        $trello->debug(1);
        $trello->fatal(1);
    
        my $board = $trello->boards('4d5ea62fd76a');
        my $results = $board->fetch;
    
        # after some introspection
    
        $board->update( ... );

DESCRIPTION

    This distribution provides an object-oriented thin-client library for
    interacting with the Trello (http://trello.com) API. For usage and
    documentation information visit
    https://trello.com/docs/gettingstarted/index.html. API::Trello is
    derived from API::Client and inherits all of it's functionality. Please
    read the documentation for API::Client for more usage information.

ATTRIBUTES

 identifier

        $trello->identifier;
        $trello->identifier('IDENTIFIER');

    The identifier attribute should be set to a string that identifies your
    application.

 key

        $trello->key;
        $trello->key('KEY');

    The key attribute should be set to the account holder's API key.

 token

        $trello->token;
        $trello->token('TOKEN');

    The token attribute should be set to the account holder's API access
    token.

 identifier

        $trello->identifier;
        $trello->identifier('IDENTIFIER');

    The identifier attribute should be set using a string to identify your
    app.

 debug

        $trello->debug;
        $trello->debug(1);

    The debug attribute if true prints HTTP requests and responses to
    standard out.

 fatal

        $trello->fatal;
        $trello->fatal(1);

    The fatal attribute if true promotes 4xx and 5xx server response codes
    to exceptions, a API::Client::Exception object.

 retries

        $trello->retries;
        $trello->retries(10);

    The retries attribute determines how many times an HTTP request should
    be retried if a 4xx or 5xx response is received. This attribute
    defaults to 0.

 timeout

        $trello->timeout;
        $trello->timeout(5);

    The timeout attribute determines how long an HTTP connection should be
    kept alive. This attribute defaults to 10.

 url

        $trello->url;
        $trello->url(Mojo::URL->new('https://api.trello.com'));

    The url attribute set the base/pre-configured URL object that will be
    used in all HTTP requests. This attribute expects a Mojo::URL object.

 user_agent

        $trello->user_agent;
        $trello->user_agent(Mojo::UserAgent->new);

    The user_agent attribute set the pre-configured UserAgent object that
    will be used in all HTTP requests. This attribute expects a
    Mojo::UserAgent object.

METHODS

 action

        my $result = $trello->action($verb, %args);
    
        # e.g.
    
        $trello->action('head', %args);    # HEAD request
        $trello->action('options', %args); # OPTIONS request
        $trello->action('patch', %args);   # PATCH request

    The action method issues a request to the API resource represented by
    the object. The first parameter will be used as the HTTP request
    method. The arguments, expected to be a list of key/value pairs, will
    be included in the request if the key is either data or query.

 create

        my $results = $trello->create(%args);
    
        # or
    
        $trello->POST(%args);

    The create method issues a POST request to the API resource represented
    by the object. The arguments, expected to be a list of key/value pairs,
    will be included in the request if the key is either data or query.

 delete

        my $results = $trello->delete(%args);
    
        # or
    
        $trello->DELETE(%args);

    The delete method issues a DELETE request to the API resource
    represented by the object. The arguments, expected to be a list of
    key/value pairs, will be included in the request if the key is either
    data or query.

 fetch

        my $results = $trello->fetch(%args);
    
        # or
    
        $trello->GET(%args);

    The fetch method issues a GET request to the API resource represented
    by the object. The arguments, expected to be a list of key/value pairs,
    will be included in the request if the key is either data or query.

 update

        my $results = $trello->update(%args);
    
        # or
    
        $trello->PUT(%args);

    The update method issues a PUT request to the API resource represented
    by the object. The arguments, expected to be a list of key/value pairs,
    will be included in the request if the key is either data or query.

RESOURCES

 actions

        $trello->actions;

    The actions method returns a new instance representative of the API
    actions resource requested. This method accepts a list of path segments
    which will be used in the HTTP request. The following documentation can
    be used to find more information.
    https://trello.com/docs/api/action/index.html.

 batch

        $trello->batch;

    The batch method returns a new instance representative of the API batch
    resource requested. This method accepts a list of path segments which
    will be used in the HTTP request. The following documentation can be
    used to find more information.
    https://trello.com/docs/api/batch/index.html.

 boards

        $trello->boards;

    The boards method returns a new instance representative of the API
    boards resource requested. This method accepts a list of path segments
    which will be used in the HTTP request. The following documentation can
    be used to find more information.
    https://trello.com/docs/api/board/index.html.

 cards

        $trello->cards;

    The cards method returns a new instance representative of the API cards
    resource requested. This method accepts a list of path segments which
    will be used in the HTTP request. The following documentation can be
    used to find more information.
    https://trello.com/docs/api/card/index.html.

 checklists

        $trello->checklists;

    The checklists method returns a new instance representative of the API
    checklists resource requested. This method accepts a list of path
    segments which will be used in the HTTP request. The following
    documentation can be used to find more information.
    https://trello.com/docs/api/checklist/index.html.

 labels

        $trello->labels;

    The labels method returns a new instance representative of the API
    labels resource requested. This method accepts a list of path segments
    which will be used in the HTTP request. The following documentation can
    be used to find more information.
    https://trello.com/docs/api/label/index.html.

 lists

        $trello->lists;

    The lists method returns a new instance representative of the API lists
    resource requested. This method accepts a list of path segments which
    will be used in the HTTP request. The following documentation can be
    used to find more information.
    https://trello.com/docs/api/list/index.html.

 members

        $trello->members;

    The members method returns a new instance representative of the API
    members resource requested. This method accepts a list of path segments
    which will be used in the HTTP request. The following documentation can
    be used to find more information.
    https://trello.com/docs/api/member/index.html.

 notifications

        $trello->notifications;

    The notifications method returns a new instance representative of the
    API notifications resource requested. This method accepts a list of
    path segments which will be used in the HTTP request. The following
    documentation can be used to find more information.
    https://trello.com/docs/api/notification/index.html.

 organizations

        $trello->organizations;

    The organizations method returns a new instance representative of the
    API organizations resource requested. This method accepts a list of
    path segments which will be used in the HTTP request. The following
    documentation can be used to find more information.
    https://trello.com/docs/api/organization/index.html.

 search

        $trello->search;

    The search method returns a new instance representative of the API
    search resource requested. This method accepts a list of path segments
    which will be used in the HTTP request. The following documentation can
    be used to find more information.
    https://trello.com/docs/api/search/index.html.

 sessions

        $trello->sessions;

    The sessions method returns a new instance representative of the API
    sessions resource requested. This method accepts a list of path
    segments which will be used in the HTTP request. The following
    documentation can be used to find more information.
    https://trello.com/docs/api/session/index.html.

 tokens

        $trello->tokens;

    The tokens method returns a new instance representative of the API
    tokens resource requested. This method accepts a list of path segments
    which will be used in the HTTP request. The following documentation can
    be used to find more information.
    https://trello.com/docs/api/token/index.html.

 types

        $trello->types;

    The types method returns a new instance representative of the API types
    resource requested. This method accepts a list of path segments which
    will be used in the HTTP request. The following documentation can be
    used to find more information.
    https://trello.com/docs/api/type/index.html.

 webhooks

        $trello->webhooks;

    The webhooks method returns a new instance representative of the API
    webhooks resource requested. This method accepts a list of path
    segments which will be used in the HTTP request. The following
    documentation can be used to find more information.
    https://trello.com/docs/api/webhook/index.html.

AUTHOR

    Al Newkirk <anewkirk@ana.io>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Al Newkirk.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Changelog for API-Trello

0.06 2015-10-11T08:07:01
 - Modified the implementation style and dependencies.

0.05 2015-09-28T13:18:23
 - v0.05
 - Updated build.
 - Updated inconsistency in documentation.
 - Updated dist.ini to generate META files.

0.04 2015-09-28T04:06:36
 - v0.04
 - Updated build.
 - Refactored to derive from API::Client.
 - Updated documentation.
 - Updated dist.ini.

0.03 2015-07-01T15:58:20
 - v0.03
 - Refactored dependencies.
 - Initial commit.

0.02 2015-05-22T12:54:37
 - v0.02
 - Initial commit.

0.01 2015-05-17T03:26:02
 - v0.01
 - Initial commit.


lib/API/Trello.pm
# ABSTRACT: Trello.com API Client
package API::Trello;

use Data::Object::Class;
use Data::Object::Signatures;

use Data::Object::Library qw(
    Str
);

extends 'API::Client';

# VERSION

our $DEFAULT_URL = "https://api.trello.com";

# ATTRIBUTES

has key => (
    is       => 'rw',
    isa      => Str,
    required => 1,
);

has token => (
    is       => 'rw',
    isa      => Str,
    required => 1,
);

# DEFAULTS

has '+casing' => (
    default  => 'camelcase',
    required => 0,
);

has '+identifier' => (
    default  => 'API::Trello (Perl)',
    required => 0,
);

has '+url' => (
    default  => $DEFAULT_URL,
    required => 0,
);

has '+version' => (
    default  => 1,
    required => 0,
);

# CONSTRUCTION

after BUILD => method {

    my $key        = $self->key;
    my $token      = $self->token;
    my $version    = $self->version;
    my $url        = $self->url;

    $url->path("/$version");
    $url->query(key => $key, $token ? (token => $token) : ());

    return $self;

};

# METHODS

method PREPARE ($ua, $tx, %args) {

    my $headers = $tx->req->headers;
    my $url     = $tx->req->url;

    # default headers
    $headers->header('Content-Type' => 'application/json');

}

method resource (@segments) {

    # build new resource instance
    my $instance = __PACKAGE__->new(
        debug      => $self->debug,
        fatal      => $self->fatal,
        retries    => $self->retries,
        timeout    => $self->timeout,
        user_agent => $self->user_agent,
        key        => $self->key,
        token      => $self->token,
        identifier => $self->identifier,
        version    => $self->version,
    );

    # resource locator
    my $url = $instance->url;

    # modify resource locator if possible
    $url->path(join '/', $self->url->path, @segments);

    # return resource instance
    return $instance;

}

1;

=encoding utf8

=head1 SYNOPSIS

    use API::Trello;

    my $trello = API::Trello->new(
        key        => 'KEY',
        token      => 'TOKEN',
        identifier => 'APPLICATION NAME',
    );

    $trello->debug(1);
    $trello->fatal(1);

    my $board = $trello->boards('4d5ea62fd76a');
    my $results = $board->fetch;

    # after some introspection

    $board->update( ... );

=head1 DESCRIPTION

This distribution provides an object-oriented thin-client library for
interacting with the Trello (L<http://trello.com>) API. For usage and
documentation information visit L<https://trello.com/docs/gettingstarted/index.html>.
API::Trello is derived from L<API::Client> and inherits all of it's
functionality. Please read the documentation for API::Client for more usage
information.

=cut

=attr identifier

    $trello->identifier;
    $trello->identifier('IDENTIFIER');

The identifier attribute should be set to a string that identifies your application.

=cut

=attr key

    $trello->key;
    $trello->key('KEY');

The key attribute should be set to the account holder's API key.

=cut

=attr token

    $trello->token;
    $trello->token('TOKEN');

The token attribute should be set to the account holder's API access token.

=cut

=attr identifier

    $trello->identifier;
    $trello->identifier('IDENTIFIER');

The identifier attribute should be set using a string to identify your app.

=cut

=attr debug

    $trello->debug;
    $trello->debug(1);

The debug attribute if true prints HTTP requests and responses to standard out.

=cut

=attr fatal

    $trello->fatal;
    $trello->fatal(1);

The fatal attribute if true promotes 4xx and 5xx server response codes to
exceptions, a L<API::Client::Exception> object.

=cut

=attr retries

    $trello->retries;
    $trello->retries(10);

The retries attribute determines how many times an HTTP request should be
retried if a 4xx or 5xx response is received. This attribute defaults to 0.

=cut

=attr timeout

    $trello->timeout;
    $trello->timeout(5);

The timeout attribute determines how long an HTTP connection should be kept
alive. This attribute defaults to 10.

=cut

=attr url

    $trello->url;
    $trello->url(Mojo::URL->new('https://api.trello.com'));

The url attribute set the base/pre-configured URL object that will be used in
all HTTP requests. This attribute expects a L<Mojo::URL> object.

=cut

=attr user_agent

    $trello->user_agent;
    $trello->user_agent(Mojo::UserAgent->new);

The user_agent attribute set the pre-configured UserAgent object that will be
used in all HTTP requests. This attribute expects a L<Mojo::UserAgent> object.

=cut

=method action

    my $result = $trello->action($verb, %args);

    # e.g.

    $trello->action('head', %args);    # HEAD request
    $trello->action('options', %args); # OPTIONS request
    $trello->action('patch', %args);   # PATCH request


The action method issues a request to the API resource represented by the
object. The first parameter will be used as the HTTP request method. The
arguments, expected to be a list of key/value pairs, will be included in the
request if the key is either C<data> or C<query>.

=cut

=method create

    my $results = $trello->create(%args);

    # or

    $trello->POST(%args);

The create method issues a C<POST> request to the API resource represented by
the object. The arguments, expected to be a list of key/value pairs, will be
included in the request if the key is either C<data> or C<query>.

=cut

=method delete

    my $results = $trello->delete(%args);

    # or

    $trello->DELETE(%args);

The delete method issues a C<DELETE> request to the API resource represented by
the object. The arguments, expected to be a list of key/value pairs, will be
included in the request if the key is either C<data> or C<query>.

=cut

=method fetch

    my $results = $trello->fetch(%args);

    # or

    $trello->GET(%args);

The fetch method issues a C<GET> request to the API resource represented by the
object. The arguments, expected to be a list of key/value pairs, will be
included in the request if the key is either C<data> or C<query>.

=cut

=method update

    my $results = $trello->update(%args);

    # or

    $trello->PUT(%args);

The update method issues a C<PUT> request to the API resource represented by
the object. The arguments, expected to be a list of key/value pairs, will be
included in the request if the key is either C<data> or C<query>.

=cut

=resource actions

    $trello->actions;

The actions method returns a new instance representative of the API
I<actions> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/action/index.html>.

=cut

=resource batch

    $trello->batch;

The batch method returns a new instance representative of the API
I<batch> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/batch/index.html>.

=cut

=resource boards

    $trello->boards;

The boards method returns a new instance representative of the API
I<boards> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/board/index.html>.

=cut

=resource cards

    $trello->cards;

The cards method returns a new instance representative of the API
I<cards> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/card/index.html>.

=cut

=resource checklists

    $trello->checklists;

The checklists method returns a new instance representative of the API
I<checklists> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/checklist/index.html>.

=cut

=resource labels

    $trello->labels;

The labels method returns a new instance representative of the API
I<labels> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/label/index.html>.

=cut

=resource lists

    $trello->lists;

The lists method returns a new instance representative of the API
I<lists> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/list/index.html>.

=cut

=resource members

    $trello->members;

The members method returns a new instance representative of the API
I<members> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/member/index.html>.

=cut

=resource notifications

    $trello->notifications;

The notifications method returns a new instance representative of the API
I<notifications> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/notification/index.html>.

=cut

=resource organizations

    $trello->organizations;

The organizations method returns a new instance representative of the API
I<organizations> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/organization/index.html>.

=cut

=resource search

    $trello->search;

The search method returns a new instance representative of the API
I<search> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/search/index.html>.

=cut

=resource sessions

    $trello->sessions;

The sessions method returns a new instance representative of the API
I<sessions> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/session/index.html>.

=cut

=resource tokens

    $trello->tokens;

The tokens method returns a new instance representative of the API
I<tokens> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/token/index.html>.

=cut

=resource types

    $trello->types;

The types method returns a new instance representative of the API
I<types> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/type/index.html>.

=cut

=resource webhooks

    $trello->webhooks;

The webhooks method returns a new instance representative of the API
I<webhooks> resource requested. This method accepts a list of path
segments which will be used in the HTTP request. The following documentation
can be used to find more information. L<https://trello.com/docs/api/webhook/index.html>.

=cut



lib/API/Trello/Type.pm
package API::Trello::Type;

use Type::Tiny;

use Type::Utils -all;
use Type::Library -base;
use Types::Standard -types;

# VERSION

our @EXPORT = qw(
    ARRAY
    CODE
    HASH
    INTEGER
    NUMBER
    OBJECT
    STRING
    UNDEF
    USERAGENT
    URL
);

declare "ARRAY",
    as Types::Standard::ArrayRef;

declare "CODE",
    as Types::Standard::CodeRef;

declare "HASH",
    as Types::Standard::HashRef;

declare "INTEGER",
    as Types::Standard::Int;

declare "NUMBER",
    as Types::Standard::Num;

declare "OBJECT",
    as Types::Standard::Object;

declare "STRING",
    as Types::Standard::Str;

declare "UNDEF",
    as Types::Standard::Undef;

sub iface_type {
    my ($name, %opts) = @_;

    if (defined $name) {
        $opts{name} = $name unless exists $opts{name};
        $opts{name} =~ s/:://g;
    }

    my @cans = ref($opts{can})  eq 'ARRAY' ? @{$opts{can}}  : $opts{can}  // ();
    my @isas = ref($opts{isa})  eq 'ARRAY' ? @{$opts{isa}}  : $opts{isa}  // ();
    my @does = ref($opts{does}) eq 'ARRAY' ? @{$opts{does}} : $opts{does} // ();

    my $code = $opts{constraint} // sub { 1 };

    $opts{constraint} = sub {
        my @args = @_;
        return if @isas and grep(not($args[0]->isa($_)),  @isas);
        return if @cans and grep(not($args[0]->can($_)),  @cans);
        return if @does and grep(not($args[0]->does($_)), @does);
        return if not $code->(@args);
        return 1;
    };

    $opts{bless}  = "Type::Tiny";
    $opts{parent} = "OBJECT" unless $opts{parent};

    { no warnings "numeric"; $opts{_caller_level}++ }

    return declare(%opts);
}

iface_type "URL",
    isa => 'Mojo::URL',
    can => [qw(scheme userinfo host port path query fragment)];

iface_type "USERAGENT",
    isa => 'Mojo::UserAgent',
    can => [qw(delete get head options patch post put)];

1;


lib/API/Trello/Client.pm
# Trello Client Super Class
package API::Trello::Client;

use API::Trello::Class;
use API::Trello::Exception;

use Mojo::UserAgent;
use Mojo::URL;

# VERSION

has debug => (
    is       => 'rw',
    isa      => Int,
    default  => 0,
);

has fatal => (
    is       => 'rw',
    isa      => Int,
    default  => 0,
);

has retries => (
    is       => 'rw',
    isa      => Int,
    default  => 0,
);

has timeout => (
    is       => 'rw',
    isa      => Int,
    default  => 10,
);

has url => (
    is       => 'ro',
    isa      => InstanceOf['Mojo::URL'],
    default  => fun { Mojo::URL->new("https://api.trello.com") },
);

has user_agent => (
    is       => 'ro',
    isa      => InstanceOf['Mojo::UserAgent'],
    default  => fun { Mojo::UserAgent->new },
);

my @methods = qw(
    DELETE
    GET
    HEAD
    OPTIONS
    PATCH
    POST
    PUT
);

around [@methods] => fun ($orig, $self, %args) {
    my $retries = $self->retries;
    my $ua      = $self->user_agent;

    # client timeouts
    $ua->max_redirects(0);
    $ua->connect_timeout($self->timeout);
    $ua->request_timeout($self->timeout);

    # request defaults
    $ua->on(start => fun ($ua, $tx) {
        $self->PREPARE($ua, $tx, %args);
    });

    # retry entry point
    RETRY:

    # execute transaction
    my $tx  = $self->$orig(%args);
    my $ok  = $tx->res->code !~ /(4|5)\d\d/;

    # fetch transaction objects
    my $req = $tx->req;
    my $res = $tx->res;

    # attempt logging where applicable
    if ($self->debug) {
        my $reqstr = $req->to_string;
        my $resstr = $res->to_string;

        $reqstr =~ s/\s*$/\n\n\n/;
        $resstr =~ s/\s*$/\n\n\n/;

        print STDOUT $reqstr;
        print STDOUT $resstr;
    }

    # retry transaction where applicable
    goto RETRY if $retries-- > 0 and not $ok;

    # throw exception if fatal is enabled
    if ($self->fatal and not $ok) {
        API::Trello::Exception->throw(
            res    => $res,
            code   => $res->code,
            method => $req->method,
            url    => $req->url->to_string,
        );
    }

    # return JSON
    return $res->json;
};

method DELETE (Str :$path = '', HashRef :$data = {}, HashRef :$query = {}) {
    my $ua  = $self->user_agent;
    my $url = $self->url->clone;

    $url->path(join '/', $url->path, $path) if $path;
    $url->query($url->query->merge(%$query)) if keys %$query;

    return $ua->delete($url, ({}, keys(%$data) ? (json => $data) : ()));
}

fun DESTROY {
    ; # Protect subclasses using AUTOLOAD
}

method GET (Str :$path = '', HashRef :$data = {}, HashRef :$query = {}) {
    my $ua  = $self->user_agent;
    my $url = $self->url->clone;

    $url->path(join '/', $url->path, $path) if $path;
    $url->query($url->query->merge(%$query)) if keys %$query;

    return $ua->get($url, ({}, keys(%$data) ? (json => $data) : ()));
}

method HEAD (Str :$path = '', HashRef :$data = {}, HashRef :$query = {}) {
    my $url = $self->url->clone;
    my $ua  = $self->user_agent;

    $url->path(join '/', $url->path, $path) if $path;
    $url->query($url->query->merge(%$query)) if keys %$query;

    return $ua->head($url, ({}, keys(%$data) ? (json => $data) : ()));
}

method OPTIONS (Str :$path = '', HashRef :$data = {}, HashRef :$query = {}) {
    my $url = $self->url->clone;
    my $ua  = $self->user_agent;

    $url->path(join '/', $url->path, $path) if $path;
    $url->query($url->query->merge(%$query)) if keys %$query;

    return $ua->options($url, ({}, keys(%$data) ? (json => $data) : ()));
}

method PATCH (Str :$path = '', HashRef :$data = {}, HashRef :$query = {}) {
    my $url = $self->url->clone;
    my $ua  = $self->user_agent;

    $url->path(join '/', $url->path, $path) if $path;
    $url->query($url->query->merge(%$query)) if keys %$query;

    return $ua->patch($url, ({}, keys(%$data) ? (json => $data) : ()));
}

method POST (Str :$path = '', HashRef :$data = {}, HashRef :$query = {}) {
    my $url = $self->url->clone;
    my $ua  = $self->user_agent;

    $url->path(join '/', $url->path, $path) if $path;
    $url->query($url->query->merge(%$query)) if keys %$query;

    return $ua->post($url, ({}, keys(%$data) ? (json => $data) : ()));
}

method PUT (Str :$path = '', HashRef :$data = {}, HashRef :$query = {}) {
    my $url = $self->url->clone;
    my $ua  = $self->user_agent;

    $url->path(join '/', $url->path, $path) if $path;
    $url->query($url->query->merge(%$query)) if keys %$query;

    return $ua->put($url, ({}, keys(%$data) ? (json => $data) : ()));
}

1;


lib/API/Trello/Exception.pm
# Trello Exception Class
package API::Trello::Exception;

use Data::Dumper ();
use Scalar::Util ();

use API::Trello::Class;

use overload '""' => 'to_string', fallback => 1;

# VERSION

has ['code', 'method', 'res', 'url'] => (
    is => 'ro',
);

method caught ($e) {
    return ! Scalar::Util::blessed($e)
        && UNIVERSAL::isa($e, $self);
}

method dumper {
    local $Data::Dumper::Terse = 1;
    return Data::Dumper::Dumper($self);
}

method rethrow {
    die $self;
}

method throw ($class: %args) {
    die $class->new(%args,
        subroutine => (caller(1))[3],
        package    => (caller(0))[0],
        file       => (caller(0))[1],
        line       => (caller(0))[2],
    );
}

method to_string {
    return sprintf "%s response received while processing request %s %s\n",
        $self->code, $self->method, $self->url;
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.037.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Trello.com API Client",
  "AUTHOR" => "Al Newkirk <anewkirk\@ana.io>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "API-Trello",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "API::Trello",
  "PREREQ_PM" => {
    "API::Client" => "0.01",
    "Data::Object" => "0.55",
    "Mojolicious" => "6.22",
    "namespace::autoclean" => "0.27"
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "API::Client" => "0.01",
  "Data::Object" => "0.55",
  "ExtUtils::MakeMaker" => 0,
  "Mojolicious" => "6.22",
  "namespace::autoclean" => "0.27"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
